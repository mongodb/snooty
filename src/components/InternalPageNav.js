import React from 'react';
import PropTypes from 'prop-types';
import Link from './Link';
import { getPageTitle } from '../utils/get-page-title';

const InternalPageNav = ({ slug, slugTitleMapping, toctreeOrder }) => {
  const slugIndex = toctreeOrder.indexOf(slug);
  const prevSlug = slugIndex > 0 ? toctreeOrder[slugIndex - 1] : null;
  const nextSlug = slugIndex < toctreeOrder.length - 1 ? toctreeOrder[slugIndex + 1] : null;
  return (
    <div id="btnv">
      {prevSlug && (
        <React.Fragment>
          <span className="btn-arrow-left">← &nbsp;</span>
          {/* TODO: Replace <a> with <Link> when back button behavior is fixed for the component.
              GitHub issue: https://github.com/gatsbyjs/gatsby/issues/8357 */}
          <Link className="btn-prev-text" to={prevSlug} title="Previous Section">
            <span>{getPageTitle(prevSlug, slugTitleMapping)}</span>
          </Link>
        </React.Fragment>
      )}
      {nextSlug && (
        <React.Fragment>
          <Link className="btn-next-text" to={nextSlug} title="Next Section">
            <span>{getPageTitle(nextSlug, slugTitleMapping)}</span>
          </Link>
          <span className="btn-arrow-right">&nbsp;→</span>
        </React.Fragment>
      )}
    </div>
  );
};

InternalPageNav.propTypes = {
  slug: PropTypes.string.isRequired,
  slugTitleMapping: PropTypes.objectOf(PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.object), PropTypes.string]))
    .isRequired,
  toctreeOrder: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default InternalPageNav;
