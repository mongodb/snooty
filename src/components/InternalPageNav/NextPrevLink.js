import React from 'react';
import { css, cx } from '@leafygreen-ui/emotion';
import Button from '@leafygreen-ui/button';
import { Body } from '@leafygreen-ui/typography';
import Icon from '@leafygreen-ui/icon';
import { palette } from '@leafygreen-ui/palette';
import { theme } from '../../theme/docsTheme';
import Link from '../Link';

const commonTextStyles = css`
  font-size: ${theme.fontSize.small};
  line-height: 20px;
`;

const nextPrevTextStyling = css`
  ${commonTextStyles}
  font-weight: 500;
`;

const nextTextStyling = css`
  text-align: end;
`;

const prevTextStyling = css`
  text-align: start;
`;

const nextPrevTitleTextStyling = css`
  ${commonTextStyles}
  color: ${palette.gray.base};
`;

const commonLinkContentContainerStyling = css`
  align-items: center;
  display: flex;
  column-gap: 14px;
`;

const nextLinkContainerStyling = css`
  ${commonLinkContentContainerStyling}
  flex-direction: row-reverse;
`;

const prevLinkContainerStyling = css`
  ${commonLinkContentContainerStyling}
  flex-direction: row;
`;

const NextPrevLink = ({ icon, direction, pageTitle, targetSlug, title }) => {
  const isNext = direction?.toLowerCase() === 'next';
  const isPrev = direction?.toLowerCase() === 'back';

  return (
    <Link to={targetSlug} title={title}>
      <div className={cx({ [nextLinkContainerStyling]: isNext, [prevLinkContainerStyling]: isPrev })}>
        <Button size="large">
          <Icon glyph={icon} />
        </Button>
        <div className={cx({ [nextTextStyling]: isNext }, { [prevTextStyling]: isPrev })}>
          <Body className={cx(nextPrevTextStyling)}>{direction}</Body>
          <Body className={cx(nextPrevTitleTextStyling)}>{pageTitle}</Body>
        </div>
      </div>
    </Link>
  );
};

export default NextPrevLink;
