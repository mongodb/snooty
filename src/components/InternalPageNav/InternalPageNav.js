import React from 'react';
import PropTypes from 'prop-types';
import { glyphs } from '@leafygreen-ui/icon';
import { css, cx } from '@leafygreen-ui/emotion';
import { theme } from '../../theme/docsTheme';
import { getPageTitle } from '../../utils/get-page-title';
import { useActiveMpTutorial } from '../MultiPageTutorials';
import { reportAnalytics } from '../../utils/report-analytics';
import NextPrevLink from './NextPrevLink';

const containerStyling = css`
  padding-bottom: 2.5em;
  width: 100%;
  display: flex;
  justify-content: space-between;
  column-gap: ${theme.size.default};
  margin-top: 88px;

  @media ${theme.screenSize.upToSmall} {
    flex-direction: column-reverse;
    row-gap: 64px;
    margin-top: 66px;
  }

  @media print {
    display: none;
  }
`;

const getActiveTutorialPage = (activeTutorial, key, linkTitle) => {
  return {
    targetSlug: activeTutorial[key].targetSlug,
    pageTitle: activeTutorial[key].pageTitle,
    linkTitle,
  };
};

const getTocPage = (targetSlug, slugTitleMapping, linkTitle) => {
  return {
    targetSlug,
    pageTitle: targetSlug ? getPageTitle(targetSlug, slugTitleMapping) : null,
    linkTitle,
  };
};

const getPrev = (activeTutorial, toctreeOrder, slugTitleMapping, slugIndex) => {
  const key = 'prev';
  if (activeTutorial?.[key]) {
    return getActiveTutorialPage(activeTutorial, key, 'Previous Step');
  }
  const prevSlug = slugIndex > 0 ? toctreeOrder[slugIndex - 1] : null;
  return getTocPage(prevSlug, slugTitleMapping, 'Previous Section');
};

const getNext = (activeTutorial, toctreeOrder, slugTitleMapping, slugIndex) => {
  const key = 'next';
  if (activeTutorial?.[key]) {
    return getActiveTutorialPage(activeTutorial, key, 'Next Step');
  }
  const nextSlug = slugIndex < toctreeOrder.length - 1 ? toctreeOrder[slugIndex + 1] : null;
  return getTocPage(nextSlug, slugTitleMapping, 'Next Section');
};

const InternalPageNav = ({ slug, slugTitleMapping, toctreeOrder }) => {
  const activeTutorial = useActiveMpTutorial();
  const slugIndex = toctreeOrder.indexOf(slug);
  const prevPage = getPrev(activeTutorial, toctreeOrder, slugTitleMapping, slugIndex);
  const nextPage = getNext(activeTutorial, toctreeOrder, slugTitleMapping, slugIndex);

  const handleClick = (direction, targetSlug) => {
    reportAnalytics('InternalPageNavClicked', {
      direction,
      targetSlug,
    });
  };

  return (
    <div className={cx(containerStyling)}>
      {prevPage?.targetSlug && (
        <NextPrevLink
          icon={glyphs.ArrowLeft.displayName}
          direction="Back"
          targetSlug={prevPage.targetSlug}
          pageTitle={prevPage.pageTitle}
          title={prevPage.linkTitle}
          onClick={handleClick}
        />
      )}
      {nextPage?.targetSlug && (
        <NextPrevLink
          icon={glyphs.ArrowRight.displayName}
          direction="Next"
          targetSlug={nextPage.targetSlug}
          pageTitle={nextPage.pageTitle}
          title={nextPage.linkTitle}
          onClick={handleClick}
        />
      )}
    </div>
  );
};

InternalPageNav.propTypes = {
  slug: PropTypes.string.isRequired,
  slugTitleMapping: PropTypes.objectOf(PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.object), PropTypes.string]))
    .isRequired,
  toctreeOrder: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default InternalPageNav;
