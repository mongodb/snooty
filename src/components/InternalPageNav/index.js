import InternalPageNav from './InternalPageNav';
import NextPrevLink from './NextPrevLink';

export { InternalPageNav, NextPrevLink };
