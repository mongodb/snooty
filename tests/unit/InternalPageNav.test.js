import React from 'react';
import { render } from 'enzyme';
import InternalPageNav from '../../src/components/InternalPageNav';
import slugTitleMapping from './data/ecosystem/slugTitleMapping.json';

const data = ['drivers/csharp', 'drivers/go', 'drivers/java'];

const renderNav = (slug) =>
  render(<InternalPageNav slug={slug} slugTitleMapping={slugTitleMapping} toctreeOrder={data} />);

it('renders a page with next and previous links correctly', () => {
  const tree = renderNav('drivers/go');
  expect(tree).toMatchSnapshot();
});

it('renders a page with no previous link correctly', () => {
  const tree = renderNav('drivers/csharp');
  expect(tree).toMatchSnapshot();
});

it('renders a page with no next link correctly', () => {
  const tree = renderNav('drivers/java');
  expect(tree).toMatchSnapshot();
});
