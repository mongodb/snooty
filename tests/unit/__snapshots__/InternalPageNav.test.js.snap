// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a page with next and previous links correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  padding-top: 2em;
  padding-bottom: 2.5em;
  display: table;
  width: 100%;
}

@media print {
  .emotion-0 {
    display: none;
  }
}

.emotion-1 {
  display: table-cell;
  vertical-align: middle;
  line-height: 28px;
  text-align: right;
}

.emotion-2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  position: relative;
  -webkit-text-decoration: none!important;
  text-decoration: none!important;
  line-height: 13px;
  outline: none;
  color: #016BF8;
}

.emotion-2:focus {
  outline: none;
}

.emotion-2>code {
  color: #016BF8;
}

.emotion-2::after {
  content: '';
  position: absolute;
  width: 100%;
  height: 2px;
  bottom: -4px;
  left: 0;
  border-radius: 2px;
}

.emotion-2:focus .emotion-2::after {
  background-color: #0498EC;
}

.emotion-2:hover::after {
  background-color: #E8EDEB;
}

.emotion-3 {
  display: table-cell;
  vertical-align: middle;
  line-height: 28px;
  text-align: left;
}

.emotion-4 {
  display: inline-block;
  width: 270px;
}

<div
    class="emotion-0"
  >
    <span
      class="emotion-1"
    >
      ←  
    </span>
    <a
      class="emotion-2 emotion-3"
      href="/drivers/csharp/"
      title="Previous Section"
    >
      <span
        class="emotion-4"
      >
        MongoDB C#/.NET Driver
      </span>
    </a>
    <a
      class="emotion-2 emotion-1"
      href="/drivers/java/"
      title="Next Section"
    >
      <span
        class="emotion-4"
      >
        Java MongoDB Driver
      </span>
    </a>
    <span
      class="emotion-3"
    >
       →
    </span>
  </div>
</DocumentFragment>
`;

exports[`renders a page with no next link correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  padding-top: 2em;
  padding-bottom: 2.5em;
  display: table;
  width: 100%;
}

@media print {
  .emotion-0 {
    display: none;
  }
}

.emotion-1 {
  display: table-cell;
  vertical-align: middle;
  line-height: 28px;
  text-align: right;
}

.emotion-2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  position: relative;
  -webkit-text-decoration: none!important;
  text-decoration: none!important;
  line-height: 13px;
  outline: none;
  color: #016BF8;
}

.emotion-2:focus {
  outline: none;
}

.emotion-2>code {
  color: #016BF8;
}

.emotion-2::after {
  content: '';
  position: absolute;
  width: 100%;
  height: 2px;
  bottom: -4px;
  left: 0;
  border-radius: 2px;
}

.emotion-2:focus .emotion-2::after {
  background-color: #0498EC;
}

.emotion-2:hover::after {
  background-color: #E8EDEB;
}

.emotion-3 {
  display: table-cell;
  vertical-align: middle;
  line-height: 28px;
  text-align: left;
}

.emotion-4 {
  display: inline-block;
  width: 270px;
}

<div
    class="emotion-0"
  >
    <span
      class="emotion-1"
    >
      ←  
    </span>
    <a
      class="emotion-2 emotion-3"
      href="/drivers/go/"
      title="Previous Section"
    >
      <span
        class="emotion-4"
      >
        MongoDB Go Driver
      </span>
    </a>
  </div>
</DocumentFragment>
`;

exports[`renders a page with no previous link correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  padding-top: 2em;
  padding-bottom: 2.5em;
  display: table;
  width: 100%;
}

@media print {
  .emotion-0 {
    display: none;
  }
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  position: relative;
  -webkit-text-decoration: none!important;
  text-decoration: none!important;
  line-height: 13px;
  outline: none;
  color: #016BF8;
}

.emotion-1:focus {
  outline: none;
}

.emotion-1>code {
  color: #016BF8;
}

.emotion-1::after {
  content: '';
  position: absolute;
  width: 100%;
  height: 2px;
  bottom: -4px;
  left: 0;
  border-radius: 2px;
}

.emotion-1:focus .emotion-1::after {
  background-color: #0498EC;
}

.emotion-1:hover::after {
  background-color: #E8EDEB;
}

.emotion-2 {
  display: table-cell;
  vertical-align: middle;
  line-height: 28px;
  text-align: right;
}

.emotion-3 {
  display: inline-block;
  width: 270px;
}

.emotion-4 {
  display: table-cell;
  vertical-align: middle;
  line-height: 28px;
  text-align: left;
}

<div
    class="emotion-0"
  >
    <a
      class="emotion-1 emotion-2"
      href="/drivers/go/"
      title="Next Section"
    >
      <span
        class="emotion-3"
      >
        MongoDB Go Driver
      </span>
    </a>
    <span
      class="emotion-4"
    >
       →
    </span>
  </div>
</DocumentFragment>
`;
